### ISSUE TITLE
The error "KaTeX parse error: Invalid size" occurred when using `\\ [`

### ISSUE THREAD

**Issue by ZigZagK** [KaTeX/KaTeX#3745]
**Describe the bug:**

When using `\\[` such as:
```
a+b=c\\
[f(x)]^2
```
The error `KaTeX parse error: Invalid size: 'f(x)' at position 9: a+b=c\\ [̲f̲(̲x̲)̲]̲^2` occurred.

**To Reproduce:**
Can be reproduced at [https://katex.org/](https://katex.org/).

**Environment (please complete the following information):**
 - KaTeX Version: 0.16.3

**Comment by ronkok** [comment]
That is not a bug. KaTeX is acting in precisely the same way that LaTeX acts with this input. When you write `\\[`,KaTeX is expecting you to specify a line skip distance, e.g., `\\[2em]`.

There are many work arounds. For instance
```
a+b=c\\
{}[f(x)]^2
```

**Comment by ZigZagK** [comment]
Thank you for replying.
However, I think it's a bit strange that `\\[` is considered to be a line skip distance when `\\` and `[` are not even in the same line.
And it can be parsed in MathJax successfully, so I still think it should be improved.
Thanks again!

**Comment by edemaine** [comment]
@ronkok I thought the same, but then I tested the following in LaTeX:
```latex
\begin{align*}
a+b=c \\
[f(x)]^2
\end{align*}
```
In fact, `\\` seems to detect the whitespace before the `[`, and not treat it as an optional argument in this case.

In fact, the following KaTeX code seems to be trying to implement exactly this:

https://github.com/KaTeX/KaTeX/blob/2c13d2631ff9e8a6df7d895d2b2194da0572981e/src/environments/array.js#L221-L223

In fact, the above example works in KaTeX!  What doesn't work is `\\` outside of an align environment. Technically that shouldn't work at all in LaTeX (can't write `$a \\ b$` for example), but we've accepted it as a kind of extension (inspired by MathJax).

It looks like the culprit is [`src/functions/cr.js`](https://github.com/KaTeX/KaTeX/blob/main/src/functions/cr.js), which uses `numOptionalArgs` to fetch the optional argument.  It does seem that `\@ifnextchar` consumes spaces in LaTeX, so the default processing of optional arguments is correct.  AMSMath's definition of `\math@cr` is just special to avoid consuming spaces, via `\new@ifnextchar`.
